examples/MetropolisWithinGibbs.md

```diff
@@ -0,0 +1,208 @@
+(\xx, \yy) \sim \operatorname{Normal}(\boldsymbol{0}, \boldsymbol{\Sigma}).
+$$
+
+The corresponding log-probability function is implemented below.
+
+```{code-cell} ipython3
+def logprob_fn(x, y, Sigma):
+    """
+    Log-pdf of ``(x, y) ~ Normal(0, Sigma)``.
+    """
+    z = jnp.concatenate([x, y])
+    return jsp.stats.multivariate_normal.logpdf(z, mean=jnp.zeros_like(z), cov=Sigma)
+
+# Specific example with x.shape == y.shape == (2,)
+Sigma = jnp.array([
+    [1., 0., .8, 0.],
+    [0., 1., 0., .8],
+    [.8, 0., 1., 0.],
+    [0., .8, 0., 1.]
+])
+
+def logprob(x, y):
+    return logprob_fn(x, y, Sigma=Sigma)
+```
+
+## MWG Sampling in BlackJAX
+
+In this case the conditional distributions $p(\xx \mid \yy)$ and $p(\yy \mid \xx)$ can be drawn from analytically (they are normal distributions).  However, for illustrative purposes we'll use an MCMC kernel to draw from each.  Specifically, we'll use an RMH kernel to draw from $p(\xx \mid \yy)$ and an HMC kernel to draw from $p(\yy \mid \xx)$.  To implement the corresponding MWG algorithm in BlackJAX, we'll write an `mwg_kernel()` for the problem which will do the following:
+
+1.  Maintain separate MCMC kernels to update each component of $p(\xx, \yy)$ while holding the other fixed.
+2.  Apply the kernel updates correctly.
+
+The issue with (2) is that each kernel update for a given MCMC `Algorithm` in BlackJAX refers to an algorithm-specific `AlgorithmState`.  For example, `RMHState` is a `typing.NamedTuple` class containing elements `position` and `log_probability`.  In our MWG sampling problem at the beginning of step $t$, `RMHState.log_probability` will consist of $\log p(\xx_{t-1}, \yy_{t-1})$.  After updating $\xx$, it will consist of $\log p(\xx_{t}, \yy_{t-1})$.  This happens automatically when we call `blackjax.mcmc.rmh.kernel()`.  However, after updating $\yy$ (via HMC), we must manually update `RMHState.log_probability` to consist of $\log p(\xx_{t}, \yy_{t})$.
+
+Two possible ways of doing this manual update are:
+
+1. Use the `blackjax.mcmc.algorithm.init()` function of the given component's MCMC algorithm to update the `AlgorithmState`.  This function has arguments `position` and `logprob_fn`.  In the case of the HMC component after obtaining $\xx_t$ but before drawing $\yy_t$, the `position` would be $\yy_{t-1}$ and the `logprob_fn` function would be $\log p(\xx_t, \cdot )$.
+
+2. The method above should work out-of-the-box for most (if not all) MCMC algorithms in BlackJAX.  However, it is not optimally efficient.  For example for the RMH update, after obtaining $\yy_{t-1}$ but before drawing $\xx_t$, the method above would calculate `RMHState.log_probability` to be $\log p(\xx_{t-1}, \yy_{t-1})$.  But we've already calculated this value from the previous HMC update of $\yy_{t-1} \sim p(\yy \mid \xx_{t-1})$.  So, we can save ourselves the cost of calculating the log-probability twice, at the expense of a deeper understanding of the low-level components of the algorithms at hand.
```

Comment:
I'm wondering if we should just do (1) and mentions this at the end? (1) always works and is simple, so it may be easier for people to grok the example if we only introduce one thing at a time.